upup/pkg/kutil/rollingupdate_cluster.go

```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+			return nil
+		} else if featureflag.DrainAndValidateRollingUpdate.Enabled() {
+
+			glog.Infof("Validating the cluster.")
+			if err = n.ValidateClusterWithRetries(rollingUpdateData, instanceGroupList, t); err != nil {
+				glog.Errorf("Error validating cluster: %s.", err)
+				return err
+			}
+
+			if err != nil && rollingUpdateData.FailOnValidate {
+				return fmt.Errorf("validation timed out while performing rolling update: %v", err)
+			}
+		}
+
+	}
+
+	return nil
+}
+
+// ValidateClusterWithRetries runs our validation methods on the K8s Cluster x times and then fails.
+func (n *CloudInstanceGroup) ValidateClusterWithRetries(rollingUpdateData *RollingUpdateCluster, instanceGroupList *api.InstanceGroupList, t time.Duration) (err error) {
+
+	for i := 0; i <= rollingUpdateData.ValidateRetries; i++ {
+		_, err = validation.ValidateCluster(rollingUpdateData.ClusterName, instanceGroupList, rollingUpdateData.K8sClient)
+		if err != nil {
+			glog.Infof("Waiting longer for kops validate to pass: %s.", err)
+			time.Sleep(t / 2)
+		} else {
+			glog.Infof("Cluster validated proceeding with next step in rolling update.")
+			break
+		}
+	}
+
+	return nil
+}
+
+// ValidateCluster runs our validation methods on the K8s Cluster.
+func (n *CloudInstanceGroup) ValidateCluster(rollingUpdateData *RollingUpdateCluster, instanceGroupList *api.InstanceGroupList) error {
+
+	_, err := validation.ValidateCluster(rollingUpdateData.ClusterName, instanceGroupList, rollingUpdateData.K8sClient)
+	if err != nil {
+		return fmt.Errorf("cluster %s does not pass validation", rollingUpdateData.ClusterName)
+	}
 
-		// TODO: Temporarily increase size of ASG?
+	return nil
+
+}
+
+// DeleteAWSInstance deletes an EC2 AWS Instance.
+func (n *CloudInstanceGroup) DeleteAWSInstance(u *CloudInstanceGroupInstance, c awsup.AWSCloud) error {
+
+	instanceID := aws.StringValue(u.ASGInstance.InstanceId)
+
+	glog.Infof("Stopping instance %q in AWS ASG %q.", instanceID, n.ASGName)
+
+	request := &ec2.TerminateInstancesInput{
+		InstanceIds: []*string{u.ASGInstance.InstanceId},
+	}
+	_, err := c.EC2().TerminateInstances(request)
+	if err != nil {
+		return fmt.Errorf("error deleting instance %q: %v", instanceID, err)
+	}
 
-		// TODO: Remove from ASG first so status is immediately updated?
+	return nil
+
+}
 
-		// TODO: Batch termination, like a rolling-update
+// DrainNode drains a K8s node.
+func (n *CloudInstanceGroup) DrainNode(u *CloudInstanceGroupInstance, rollingUpdateData *RollingUpdateCluster) error {
+	drain, err := NewDrainOptions(nil, rollingUpdateData.ClusterName)
+
+	if err != nil {
 
-		request := &ec2.TerminateInstancesInput{
-			InstanceIds: []*string{u.ASGInstance.InstanceId},
+		glog.Warningf("Error creating drain: %v.", err)
+		if rollingUpdateData.FailOnDrainError {
+			return fmt.Errorf("error creating drain: %v.", err)
+		} else {
+			glog.Infof("Proceeding with rolling-update since fail-on-drain-error is set to false.")
 		}
```

Comment:
The else statement is just doing logging.  I can chain the if statements together, but I personally I mess up and don't like reading long if statements.